rclrs/src/logging.rs

```diff
@@ -0,0 +1,604 @@
+    // formatting parameters in the simple message case, e.g. to write
+    // ```
+    // log_error!(<logger>, "Log with no params"); // OR
+    // log_error!(<logger>, "Log with useful info {}", error_reason);
+    ($to_log_params: expr, $($args:tt)*) => {{
+        // Adding these use statements here due an issue like this one:
+        // https://github.com/intellij-rust/intellij-rust/issues/9853
+        // Note: that issue appears to be specific to jetbrains intellisense however,
+        // observed same/similar behaviour with rust-analyzer/rustc
+        use std::sync::{Once, OnceLock, Mutex};
+        use std::time::SystemTime;
+
+        // We wrap the functional body of the macro inside of a closure which
+        // we immediately trigger. This allows us to use `return` to exit the
+        // macro early without causing the calling function to also try to
+        // return early.
+        (|| {
+            let params = $crate::ToLogParams::to_log_params($to_log_params);
+
+            if !params.get_user_filter() {
+                // The user filter is telling us not to log this time, so exit
+                // before doing anything else.
+                return;
+            }
+
+            let mut first_time = false;
+            static REMEMBER_FIRST_TIME: Once = Once::new();
+            REMEMBER_FIRST_TIME.call_once(|| first_time = true);
+
+            let logger_name = params.get_logger_name();
+            let severity = params.get_severity();
+
+            match params.get_occurence() {
+                // Create the static variables here so we get a per-instance static
+                $crate::LogOccurrence::Once => {
+                    if first_time {
+                        $crate::log_unconditional!(severity, logger_name, $($args)*);
+                    }
+                    // Since we've already logged once, we should never log again,
+                    // so just exit right now.
+                    return;
+                }
+                $crate::LogOccurrence::SkipFirst => {
+                    if first_time {
+                        // This is the first time that we're seeing this log, and we
+                        // were told to skip the first one, so just exit right away.
+                        return;
+                    }
+                }
+                // Do nothing
+                $crate::LogOccurrence::All => (),
+            }
+
+            // If we have a throttle duration then check if we're inside or outside
+            // of that interval.
+            let throttle = params.get_throttle();
+            if throttle > std::time::Duration::ZERO {
+                static LAST_LOG_TIME: OnceLock<Mutex<SystemTime>> = OnceLock::new();
+                let last_log_time = LAST_LOG_TIME.get_or_init(|| {
+                    Mutex::new(std::time::SystemTime::now())
+                });
```

Comment:
I'm not too sure about this implementation, from what I see it hardcodes the time to a `std::time::SystemTime` but in ROS 2 the clock could be a ROS time.
The `rclcpp` implementation also seems to allow specifying arbitrary clocks and has [unit tests](https://github.com/ros2/rclcpp/blob/88ebea94e9c79499c34f825cf6565827c5ba958e/rclcpp/test/rclcpp/test_logging.cpp#L178-L230) to verify that passing a ROS clock correctly throttles it.
I would suggest either making the clock to be used configurable or maybe easier just defaulting to the node's clock, so at least when running on simulated time we will behave consistently regardless of the speed of the simulation.

Reply:
Interesting, the idea of syncing it against ROS time didn't even occur to me. That strikes me as a weird use case, but I'm not opposed to it.

We can't rely on the node clock without making it mandatory to pass in a reference to the node. That could make logging unusable in code regions where it shouldn't have to be.

I don't mind making this configurable by adding a `.throttle_clock( _ )` option to `LogParams` that takes in an enum like

```
pub enum ThrottleClock {
    Clock(Clock),
    SystemTime,
    SteadyTime,
}
```

and this would default to `SteadyTime` (which I think is what 99.99% of use cases would want).

Reply:
Added in https://github.com/geoff-imdex/ros2_rust/pull/13 and just waiting to be merged in that PR

Reply:
I admit I have never used the throttling API but when I'm in doubt on "what is the usage in the ecosystem" I usually just throw a [global github search](https://github.com/search?q=RCLCPP_DEBUG_THROTTLE&type=code) and glance at the code out there.

It is probably mandated by the fact that the macro in C++ requires you to pass a clock and it is probably easier to pass the node clock rather than create a new steady_clock just for that purpose, but it seems the majority of applications do indeed pass in the node clock (which would be running in either system time or simulated time depending on parameters).
I also wonder, what's the reason for redefining a `ThrottleClock` enum rather than using [Clock](https://github.com/ros2-rust/ros2_rust/blob/066dd7c6e70d515268945ab6190793ef605fd887/rclrs/src/clock.rs#L43-L58)? It also offers steady time / system time enum variants

Reply:
>  it seems the majority of applications do indeed pass in the node clock

Are you suggesting that we make it mandatory, (e.g. instead of `.throttle(duration)` we only offer `.throttle(clock, duration)`)? If so, what do you perceive as the benefit of mandating an explicit clock instead of using steady clock by default and letting the user override that?

In my mind ROS time exists primarily to make sure that system behavior can be consistent across real world usage and simulated usage. Log throttling doesn't (..or at least shouldn't..) have an effect on the behavior of the system beyond preventing log spam. It's very hard for me to imagine a case where using ROS time is preferable for reducing log spam, so I would much prefer minimizing API noise and providing a sensible default which will probably always be what users actually want.

In any case the throttle time source type [is an optional parameter in `rclpy`'s logging API](https://github.com/ros2/rclpy/blob/34f9e13e72ea4528bd96f1d739bc7a8711cc923b/rclpy/rclpy/impl/rcutils_logger.py#L350-L354) with a default of steady time, which is exactly what I've done in https://github.com/geoff-imdex/ros2_rust/pull/13, so we'll be matching the API of rclpy even if we don't match the API of rclcpp.

> what's the reason for redefining a ThrottleClock enum rather than using [Clock](https://github.com/ros2-rust/ros2_rust/blob/066dd7c6e70d515268945ab6190793ef605fd887/rclrs/src/clock.rs#L43-L58)? It also offers steady time / system time enum variants

The difference is `ThrottleClock` borrows a clock so there's no need to clone a clock. Cloning a clock would involve incrementing and decrementing an atomic int, which isn't a terrible amount of overhead but it does involve cache syncs which should be avoided whenever possible for hot paths.

Reply:
> Are you suggesting that we make it mandatory, (e.g. instead of .throttle(duration) we only offer .throttle(clock, duration))? If so, what do you perceive as the benefit of mandating an explicit clock instead of using steady clock by default and letting the user override that?

Not really, I see the benefit of the sensible default, which is why I mentioned that probably people pass in the node clock just because the API forces them to and that's the most easily available clock. I'm fine with the current API.

> The difference is ThrottleClock borrows a clock so there's no need to clone a clock. Cloning a clock would involve incrementing and decrementing an atomic int, which isn't a terrible amount of overhead but it does involve cache syncs which should be avoided whenever possible for hot paths.

I see, my main point was about the underlying implementation as well, I am not quite sure but would it make sense to use a `Clock` with `ClockKind::SystemTime` / `ClockKind::SteadyTime` for the underlying implementation in the macro, rather than relying on Rust builtin methods?
I can see how the Rust method is probably cleaner and faster since it doesn't need to build an rcl clock or go through bindings. I however do not honestly know if the behavior of the two is guaranteed to be the same or not, and whether that might be an issue (I suppose both will do the same underlying system calls to get a steady / system time so it's probably OK?).
I'm also slightly put off by having different ways to deal with time internally (rcl / rust std), but I suppose this being purely an implementation detail and not exposed to users it is probably not a big deal?

Reply:
> I can see how the Rust method is probably cleaner and faster since it doesn't need to build an rcl clock or go through bindings.

Yes, all of this, plus the fact that going through the rcl bindings also forces a heap allocation (creating the `Arc<Mutex<rcl_clock_t>>`) which is something to avoid for hot paths, and the whole point of throttling a log is because that log exists in a hot path.

I do agree that there's a somewhat silly amount of duplication of clock abstractions going on here, but here's the way I would look at it:
* `ThrottleClock` allows a choice of clock to be made at the point in the code where the logging is happening
* `ThrottleClock::Clock` allows a choice of clock to be made elsewhere, and the logging will just abide by that choice

> I suppose both will do the same underlying system calls to get a steady / system time so it's probably OK?

I would expect so as well. I've very surprised if they're using different syscalls.

> I'm also slightly put off by having different ways to deal with time internally (rcl / rust std)

Yeah I don't love it myself. It mostly comes from `rclrs` trying to accommodate the full breadth of `rcl`'s clock interface. Personally I wouldn't mind if we narrow down on what we support from the API of `rcl_clock_t`.. maybe even not use `rcl_clock_t` at all and just implement the whole abstract clock concept on the Rust side without any rcl baggage..? But that's definitely out of scope for this PR.